### DISCUSSION TITLE
Failed to override styles when selectors are grouped

### DISCUSSION THREAD

**Discussion by Yixian15** [microsoft/griffel#495] [Q&A] [answered]
Hi Griffel developers,

I ran into an issue when trying to override CSS styles with mergeStyles. After some investigation I think it might relate to selctor matching issue.

The original style is defined with grouped selectors ".active,:hover":

```
square: {
    width: "100px",
    height: "100px",
    ...shorthands.border("2px", "solid", "black"),
    ":active,:hover": {
      backgroundColor: "blue",
    },
  },
```

Then tried to override the behavior of ":hover":

```
 hoverOverride1: {
    ":hover": {
      backgroundColor: "red",
    },
  },
```

This is how the styles are used:

```
<div className={mergeClasses(styles.square, styles.hoverOverride1)}>
  hover me
</div>
```

It is expected to override the previous definition by the later one. But it doesn't work.

Then I tried other ways to override the styles. It turns out override only works when the exact same grouped selectors are used. Even one whitespace between the selectors can fail it.

For example:

```
  hoverOverride2: {
    ":active, :hover": { // This won't work due to one extra space between the selectors
      backgroundColor: "red",
    },
  },
  hoverOverride3: {
    ":active,:hover": { // This works because the selectors are identical to the original definition
      backgroundColor: "red",
    },
  },

```

You can see the demo in the [sandbox](https://codesandbox.io/p/sandbox/griffel-pseudo-style-hzh6s5?file=%2Fsrc%2FApp.js%3A30%2C30&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrix3a3900063j6fxf5swmi9%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrix3a3900023j6f8nq95i0p%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrix3a3900033j6furf0837g%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrix3a3900053j6fn0m52wzo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrix3a3900023j6f8nq95i0p%2522%253A%257B%2522id%2522%253A%2522clrix3a3900023j6f8nq95i0p%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrixq7ep002x3j6fcdjr1o5u%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A30%252C%2522startColumn%2522%253A30%252C%2522endLineNumber%2522%253A30%252C%2522endColumn%2522%253A30%257D%255D%257D%255D%252C%2522activeTabId%2522%253A%2522clrixq7ep002x3j6fcdjr1o5u%2522%257D%252C%2522clrix3a3900053j6fn0m52wzo%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrix3a3900043j6f31bpd8hy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clrix3a3900053j6fn0m52wzo%2522%252C%2522activeTabId%2522%253A%2522clrix3a3900043j6f31bpd8hy%2522%257D%252C%2522clrix3a3900033j6furf0837g%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clrix3a3900033j6furf0837g%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by layershifter** [comment]
Hey @Yixian15. Thanks for bringing this up 👍 

Indeed Griffel is "picky" about selectors (or keys on an object). `:hover,:active` & `:hover` are considered as different rules and can't be deduplicated properly:

```ts
makeStyles({
  root: {
    ":hover": { color: "red" },
    ":hover,:active": { color: "blue" },
  },
});
```

[(playground link)](https://griffel.js.org/try-it-out/?code=JYWwDg9gTgLgBAbziAhgawKYGUYE8A2GAzgDRxEAW0MFKAdgCZFwC%2BcAZlBCHAOQACAcyjB27DPgD0AY2gZeAbgBQSjAA9IsOAwzsUAV3zxUmHAWIAKBErhwuEGAC5EN23ABEjqgDcMUT4hwsvjQzu5QGAzurCSutp4%2BfiSOKNIwwL4BSMGhHgBG%2BPoY0SyxtqVKLACUykA)

---

That's not limited to this case, as it's true for any defined selector... To avoid the problem, in the case that you mentioned it should be declared like:

```ts
const sharedStyles = { color: "red" }

makeStyles({
  root: {
    ":hover": sharedStyles,
    ":active": sharedStyles
  },
});
```

I am not sure that anything can be done there to make it more predictable as this problem is a part of design...

We can emit a warning in dev environment and add a ESLint rule to avoid such selectors as the change in transform logic does not seem to be something easy... WDYT?

**Reply by Yixian15** [reply]
Here's the case I'm working on. We're trying override styles of fluent components in our application.

The styles are defined here in fluent: https://github.com/microsoft/fluentui/blob/fcbadc1a22fe17a3c0a96bf88746e25e7def19a8/packages/react-components/react-input/src/components/Input/useInputStyles.styles.ts#L122

There's no way to override styles of ':focus-within' without affecting the styles of ':active', right? I think this could be a problem.

**Reply by layershifter** [reply, marked as answer]
> There's no way to override styles of ':focus-within' without affecting the styles of ':active', right? I think this could be a problem.

You are correct. There is no predictable way to override it. I created #499 to introduce a lint rule.

---

Now you can do a nasty hack to increase specificity:

```ts
const useClasses = makeStyles({
  c: {
    ':hover&': {
      color: 'blue',
    },
  },
});

// ⬇️⬇️⬇️
// .fms2va0:hover.fms2va0 { color: blue; }
```

https://stackblitz.com/edit/stackblitz-starters-2xtajv
